tensorflow/python/keras/layers/preprocessing/preprocessing_stage_functional.py

```diff
@@ -0,0 +1,141 @@
+    """
+    if not isinstance(data, dataset_ops.Dataset):
+      data = self._flatten_to_reference_inputs(data)
+      if any([not isinstance(datum, (np.ndarray, ops.EagerTensor))
+              for datum in data]):
+        raise ValueError(
+            '`adapt()` requires a batched Dataset, a list of EagerTensors '
+            'or Numpy arrays as input, got {}'.format(type(data)))
+
+    if isinstance(data, dataset_ops.Dataset):
+      # Validate the datasets to try and ensure we haven't been passed one with
+      # infinite size. That would cause an infinite loop here.
+      if tf_utils.dataset_is_infinite(data):
+        raise ValueError(
+            'The dataset passed to `adapt()` has an infinite number of '
+            'elements. Please use dataset.take(...) to make the number '
+            'of elements finite.')
+      # Unzip dataset object to a list of single input dataset.
+      data = _unzip_dataset(data)
+
+    # Dictionary mapping reference tensors to data
+    data_dict = {}
+    tensor_usage_count = self._tensor_usage_count
+    for x, y in zip(self.inputs, data):
+      if isinstance(y, ops.EagerTensor):
+        y = self._conform_to_reference_input(y, ref_input=x)
+      x_id = str(id(x))
+      data_dict[x_id] = [y] * tensor_usage_count[x_id]
+
+    nodes_by_depth = self._nodes_by_depth
+    depth_keys = sorted(nodes_by_depth.keys(), reverse=True)
+    for depth in depth_keys:
+      for node in nodes_by_depth[depth]:
+        # Input node
+        if node.is_input:
+          continue
+
+        # Node with input not computed yet
+        if any(t_id not in data_dict for t_id in node.flat_input_ids):
+          continue
+
+        args, kwargs = node.map_arguments(data_dict)
+
+        if hasattr(node.layer, 'adapt'):
+          args_flatten = nest.flatten(args)
+          if all(isinstance(arg, dataset_ops.Dataset) for arg in args_flatten):
+            args = dataset_ops.Dataset.zip(tuple(args_flatten))
+          node.layer.adapt(args, reset_state=reset_state)
+
+        if isinstance(args, dataset_ops.Dataset):
+          def map_fn(*x, node=node, args=args, kwargs=kwargs):
+            if len(args.element_spec) == 1:
+              return nest.flatten(node.layer(*x, **kwargs))
+            return nest.flatten(node.layer(x, **kwargs))
+          outputs = args.map(map_fn)
+          outputs = _unzip_dataset(outputs)
+        else:
+          outputs = node.layer(*args, **kwargs)
+          outputs = nest.flatten(outputs)
+
+        # Update tensor_dict.
+        for x_id, y in zip(node.flat_output_ids, outputs):
+          data_dict[x_id] = [y] * tensor_usage_count[x_id]
+
+
+def _unzip_dataset(data):
+  """Unzip dataset into a list of single element datasets.
+  Arguments:
+    data: A Dataset object.
+  Return:
+    A list of Dataset object, each correspond to one of the
+    `element_spec` of the input Dataset object.
+  Example:
+    >>> ds1 = tf.data.Dataset.from_tensor_slices([1,2,3])
```

Comment:
- add a space after the commas
- unindent the >>>
- Add a line break before each section (e.g. Returns:, Example:)